dao-voting/src/main.sw

```diff
@@ -0,0 +1,327 @@
+        });
+    }
+
+    /// Update the user balance to indicate they have withdrawn governance tokens
+    ///
+    /// # Arguments
+    ///
+    /// - `amount` - amount of governance tokens to withdraw from the contract
+    ///
+    /// # Reverts
+    ///
+    /// * When the user tries to withdraw 0 from their balance
+    /// * When the user tries to withdraw more than their balance
+    #[storage(read, write)]fn withdraw(amount: u64) {
+        require(0 < amount, UserError::AmountCannotBeZero);
+        let user: Identity = msg_sender().unwrap();
+
+        let prev_balance = storage.balances.get(user);
+        require(amount <= prev_balance, UserError::InsufficientBalance);
+
+        storage.balances.insert(user, prev_balance - amount);
+
+        // Transfer the asset back to the user
+        transfer(amount, storage.token, user);
+
+        log(WithdrawEvent {
+            amount, user: user, 
+        })
+    }
+
+    /// Vote on a given proposal
+    ///
+    /// # Arguments
+    ///
+    /// - `approve` - whether the user voted yes or no on the proposal
+    /// - `proposal_id` - Identifier used to specifiy a proposal (0 <= proposal_id < proposal_count)
+    /// - `vote_amount` - the amount of votes to cast on the proposal
+    ///
+    /// # Reverts
+    ///
+    /// * When the given proposal id is greater than or equal to proposal_count
+    /// * When the vote amount is 0
+    /// * When the proposal has passed its deadline
+    /// * When the vote amount is greater than the user's deposited balance
+    #[storage(read, write)]fn vote(approve: bool, proposal_id: u64, vote_amount: u64) {
+        validate_id(proposal_id, storage.proposal_count);
+        require(0 < vote_amount, UserError::VoteAmountCannotBeZero);
+
+        let mut proposal = storage.proposals.get(proposal_id);
+        require(height() <= proposal.deadline, ProposalError::ProposalExpired);
+
+        let user = msg_sender().unwrap();
+        let user_balance = storage.balances.get(user);
+
+        require(vote_amount <= user_balance, UserError::InsufficientBalance);
+
+        let mut votes = storage.votes.get((user, proposal_id));
+        if approve {
+            proposal.yes_votes += vote_amount;
+            votes.yes_votes += vote_amount;
+        } else {
+            proposal.no_votes += vote_amount;
+            votes.no_votes += vote_amount;
+        };
+
+        storage.balances.insert(user, user_balance - vote_amount);
+        storage.votes.insert((user, proposal_id), votes);
+        storage.proposals.insert(proposal_id, proposal);
+
+        log(VoteEvent {
+            id: proposal_id, user, vote_amount
+        });
+    }
+
+    /// Execute a given proposal
+    ///
+    /// # Arguments
+    ///
+    /// - `proposal_id` - Identifier used to specifiy a proposal (0 <= proposal_id < proposal_count)
+    ///
+    /// # Reverts
+    ///
+    /// * When the given proposal id is greater than or equal to proposal_count
```

Comment:
```suggestion
    /// * When the given proposal id is greater than or equal to proposal_count
    /// * When the proposal has already been executed
```

Reply:
I don't understand this suggestion could you elaborate?

Reply:
Revert documenting the addition of

https://github.com/FuelLabs/sway-applications/blob/5a578e5b1b55d0cd43bba988aa261bfb3ae7c078/dao-voting/src/main.sw#L217

Reply:
lol my b I've been looking at this code too long...